lms/djangoapps/onboarding/views.py

```diff
@@ -566,39 +507,68 @@ def admin_activation(request, org_id, activation_key):
 
     """
     context = {}
-    hash_key_obj = None
+    admin_activation = eval(request.GET.get('admin_activation', 'False'))
 
     try:
-        hash_key_obj = OrganizationAdminHashKeys.objects.get(activation_hash=activation_key)
-        user_extended_profile = UserExtendedProfile.objects.get(user__email=hash_key_obj.suggested_admin_email)
+        hash_key = OrganizationAdminHashKeys.objects.get(activation_hash=activation_key)
+        admin_change_confirmation = True if eval(request.GET.get('confirm', u'False')) else False
+        user_profile = UserExtendedProfile.objects.get(user__email=hash_key.suggested_admin_email)
+        user = user_profile.user
 
-        context['key'] = hash_key_obj.activation_hash
-        if hash_key_obj.is_hash_consumed:
+        context['key'] = hash_key.activation_hash
+        if hash_key.is_hash_consumed:
             activation_status = 2
         else:
+            hash_key.is_hash_consumed = True
+            hash_key.save()
             activation_status = 4
 
-        if request.method == "POST":
-            hash_key_obj.organization.admin = user_extended_profile.user
-            hash_key_obj.organization.unclaimed_org_admin_email = None
-            hash_key_obj.organization.save()
-            activation_status = 1
+        if request.method == 'POST':
+            if admin_activation:
+                hash_key.organization.unclaimed_org_admin_email = None
+                hash_key.organization.admin = user
+                hash_key.organization.save()
+                activation_status = 1
+            else:
+                if admin_change_confirmation:
+                    if user_profile.organization.admin == user:
+                        user_profile.organization.admin = None
+                        user_profile.organization.save()
+
+                    hash_key.organization.unclaimed_org_admin_email = None
+                    hash_key.organization.admin = user
+                    hash_key.organization.save()
+
+                    user_profile.organization = hash_key.organization
```

Comment:
remove duplicate code from 527 - 540
+ why are you assigning organization from has_key object?